### PR TITLE
fixed a bug in the smallestCut method

### PR DESCRIPTION
**Pull request by Felalolf**
There was a bug in the smallestCut method causing the result to not be the smallest cut. I fixed this.
Furthermore, I changed the default parameter of the penaltiy function.

### CODE REVIEW

**Review by jcp19 — APPROVED** [review]
Looks fine to me, I just have some minor comments

File: src/main/scala/viper/gobra/util/ViperChopper.scala [hunk 1]
```diff
@@ -71,7 +71,7 @@ object ViperChopper {
       val result = {
         if (nodes.size < 10) {
           val t1 = System.nanoTime()
-          val smallestPrograms = smallestCut(N, nodes, edges, (x: Int) => x)
+          val smallestPrograms = smallestCutWithCycles(N, nodes, edges, (x: Int) => x)
```
[the remark this answers is not in the document]
Comment (Felalolf) [reply src/main/scala/viper/gobra/util/ViperChopper.scala]: put `indentity[Int]`

File: src/main/scala/viper/gobra/util/ViperChopper.scala [hunk 2]
```diff
@@ -124,88 +124,113 @@ object ViperChopper {
...
 
-      // Stores whether a note was visited
-      val visited  = Array.ofDim[Boolean](N)
+      // Stores whether a node is not a root.
+      val notRoot = Array.ofDim[Boolean](N)
 
-      // Stores whether a color is not a root. Only valid if color was used.
-      val colorIsNotRoot  = Array.ofDim[Boolean](N)
+      // Stores all dependencies of a node (including itself).
+      // Serves as memorization table.
```
[the remark this answers is not in the document]
Comment (Felalolf) [reply src/main/scala/viper/gobra/util/ViperChopper.scala]: yes

File: src/main/scala/viper/gobra/util/ViperChopper.scala [hunk 3]
```diff
@@ -124,88 +124,113 @@ object ViperChopper {
...
-        * This is done by applying dfs twice. The first run precomputes results that make the second run faster.
-        * The first run colors all nodes and computes the graph induced by the coloring,
-        * resulting in a smaller graph for the second run. All members of a color are sorted immediately to optimize
-        * intermediate results for the second run.
         */
 
-      // Stores the color of nodes. A color of a node is only valid if the node was visited.
+      // Stores color of a node.
+      // Color 0 means that the node was never visited.
+      // Color 1 means that the node was visited, but not yet finalized.
```
[the remark this answers is not in the document]
Comment (Felalolf) [reply src/main/scala/viper/gobra/util/ViperChopper.scala]: I think the impact on performance is fine. On can also have two arrays one with the number of visits and one with the color. I implemented your idea, but I named the classes differently.

**Pull request merged**